core/src/main/java/org/commonjava/indy/core/inject/CassandraNotFoundCache.java

```diff
@@ -0,0 +1,273 @@
+
+        preparedQueryByStore = session.prepare( "SELECT * FROM " + keyspace + ".nfc WHERE storekey=?;" );
+
+        preparedDeleteByStore = session.prepare( "DELETE FROM " + keyspace + ".nfc WHERE storekey=?;" );
+
+        preparedDelete = session.prepare( "DELETE FROM " + keyspace + ".nfc WHERE storekey=? AND path=?;" );
+
+        preparedInsert = session.prepare( "INSERT INTO " + keyspace
+                                                          + ".nfc (storekey,path,creation,expiration) VALUES (?,?,?,?) USING TTL ?;" ); // ttl in seconds
+    }
+
+    @Override
+    protected IndyConfiguration getIndyConfiguration()
+    {
+        return config;
+    }
+
+    @Override
+    @Measure
+    public void addMissing( final ConcreteResource resource )
+    {
+        KeyedLocation location = (KeyedLocation) resource.getLocation();
+        StoreKey key = location.getKey();
+
+        int timeoutInSeconds = DEFAULT_NOT_FOUND_CACHE_TIMEOUT_SECONDS;
+        int t = getTimeoutInSeconds( resource );
+        if ( t > 0 )
+        {
+            timeoutInSeconds = t;
+        }
+        Date curDate = new Date();
+        Date timeoutDate = new Date( curDate.getTime() + ( timeoutInSeconds * 1000 ) );
+        logger.debug( "[NFC] {} will not be checked again until {}", resource,
+                      new SimpleDateFormat( TIMEOUT_FORMAT ).format( timeoutDate ) );
+
+        BoundStatement bound = preparedInsert.bind( key.toString(), resource.getPath(), curDate, timeoutDate,
+                                                    timeoutInSeconds );
+        session.execute( bound );
+    }
+
+    @Override
+    @Measure
+    public boolean isMissing( final ConcreteResource resource )
+    {
+        StoreKey key = getResourceKey( resource );
+        BoundStatement bound = preparedExistQuery.bind( key.toString(), resource.getPath() );
+        ResultSet result = session.execute( bound );
+        long count = result.one().get( 0, Long.class );
+        boolean missing = count > 0;
+        logger.trace( "NFC check: {}, missing: {}", resource, missing );
+        return missing;
+    }
+
+    @Override
+    @Measure
+    public void clearMissing( final Location location )
+    {
+        StoreKey key = ( (KeyedLocation) location ).getKey();
+        BoundStatement bound = preparedDeleteByStore.bind( key.toString() );
+        session.execute( bound );
+    }
+
+    @Override
+    @Measure
+    public void clearMissing( final ConcreteResource resource )
+    {
+        StoreKey key = getResourceKey( resource );
+        BoundStatement bound = preparedDelete.bind( key.toString(), resource.getPath() );
+        session.execute( bound );
+    }
+
+    @Override
+    @Measure
+    public void clearAllMissing()
+    {
+        session.execute( "TRUNCATE " + keyspace + ".nfc;" );
+    }
+
+    @Override
+    @Measure
+    public Map<Location, Set<String>> getAllMissing()
+    {
+        return Collections.emptyMap(); // not support
```

Comment:
Is this because we can't execute a query that spans partition keys? Or is there a concern that this will be too expensive somehow?

Reply:
The entries span multiple partitions. We can force a query to select all but that will be slow or very slow (the more the cluster nodes, the slower).